pkg/apis/serving/v1alpha1/revision_validation.go

```diff
@@ -133,6 +134,28 @@ func validateContainer(container corev1.Container) *apis.FieldError {
 	return errs
 }
 
+func validateContainerPorts(ports []corev1.ContainerPort) *apis.FieldError {
+	// user can set container port which names "user-port" to define application's port.
+	// Queue-proxy will use it to send requests to application
+	// if user didn't set any port, it will set default port user-port=8080.
+	if len(ports) > 1 {
+		return &apis.FieldError{
+			Message: "More than one container ports is set",
+			Paths:   []string{apis.CurrentField},
+			Details: "Only a single port named \"user-port\" is allowed",
+		}
+	}
+
+	if len(ports) == 1 && ports[0].Name != "user-port" {
```

Comment:
so it means accept any name of the port ? 
I get it.

Reply:
To clarify my understanding of the contract, there should only ever be 1 port in the Configuration Spec (or configuration section of Service Spec). Configuration.Spec.Container.Ports[0].Name should either be "" or "http1" or "h2c". Since we currently only implement automatic detection it probably makes sense to only accept "" for Name.

When the revision is created by Knative, the revision creates a K8s Deployment object from the spec. The port name on the Deployment is currently being set to 'user-port', and I think it makes sense to keep this set for clarity that the 'user-port' containerPort on the deployment was set from Configuration.Spec.Container.Ports[0].

Reply:
To quote the spec (because apparently I had deep discussions about this 6 months ago, and was smarter than I am when looking at this issue for 10 minutes today):

> Only one inbound `containerPort` SHALL be specified in the [`core.v1.Container`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.12/#containerport-v1-core) specification. The `hostPort` parameter SHOULD NOT be set by the developer or the platform provider, as it can interfere with ingress autoscaling. Regardless of its source, the selected port will be made available in the `PORT` environment variable.
> 
> The platform provider SHOULD configure the platform to perform HTTPS termination and protocol transformation e.g. between QUIC or HTTP/2 and HTTP/1.1. Developers should not need to implement multiple transports between the platform and their code. **Unless overridden by setting the `name` field on the inbound port**, the platform will perform automatic detection as described above. If the `core.v1.Container.ports[0].name` is set to one of the following values, HTTP negotiation will be disabled and the following protocol will be used:
> 
> * `http1`: HTTP/1.1 transport and will not attempt to upgrade to h2c..
> * `h2c:` HTTP/2 transport, as described in section 3.4 of the HTTP2 spec (Starting HTTP/2 with Prior Knowledge)
>
> Developers SHOULD prefer to use automatic content negotiation where available, and **MUST NOT set the name field to arbitrary values, as additional transports may be defined in the future.**

All of the above is about the Container object specified in the Revision. The underlying implementation (e.g. creating a Deployment to back the Revision) can use whatever names it wants. The content-negotiation information needs to be forwarded to queue-proxy (this is not done today).